### DISCUSSION TITLE
Anyone run into the linux docker base build failing at step 30 `RUN gclient sync` with cipd failing with `invalid cross-device link`?

### DISCUSSION THREAD

**Discussion by josh-hemphill** [caoccao/Javet#196] [Q&A]
Not sure yet if this warrants an issue here or in the chromium-dev groups. I assume others can run and build the docker containers just fine?
I think I've eliminated it being my host machine, I've run it on 2 different physical machines, a vm, and tried with 2 different container backends.

I actually also run into the `RUN fetch v8` failing with `tar` permissions, but I can set the tar option `--no-same-owner` and that goes along fine but fails at `gclient sync`. I can also adjust the file to run as a non-root user which solves the permissions issue and everything runs up until the exact same error.

I was going to work on a build for linux arm64, but this happens in the base build whether I change things or not.

The exact error is:
```
STEP 30/71: RUN gclient sync
Updating depot_tools...

________ running 'cipd ensure -log-level error -root /google -ensure-file /tmp/tmpb5jmmqu3.ensure' in '.'
[P730 08:39:52.369 client.go:1901 E] [cleanup] Failed to remove infra/3pp/tools/ninja/linux-amd64 in "v8/third_party/ninja": removing the deployed package directory: rename /google/.cipd/pkgs/1 /google/.cipd/pkgs/B9dhvBqFSHhs: invalid cross-device link
Errors:
  failed to remove infra/3pp/tools/ninja/linux-amd64 in "v8/third_party/ninja": removing the deployed package directory: rename /google/.cipd/pkgs/1 /google/.cipd/pkgs/B9dhvBqFSHhs: invalid cross-device link
Error: Command 'cipd ensure -log-level error -root /google -ensure-file /tmp/tmpb5jmmqu3.ensure' returned non-zero exit status 1
[P730 08:39:52.369 client.go:1901 E] [cleanup] Failed to remove infra/3pp/tools/ninja/linux-amd64 in "v8/third_party/ninja": removing the deployed package directory: rename /google/.cipd/pkgs/1 /google/.cipd/pkgs/B9dhvBqFSHhs: invalid cross-device link
Errors:
  failed to remove infra/3pp/tools/ninja/linux-amd64 in "v8/third_party/ninja": removing the deployed package directory: rename /google/.cipd/pkgs/1 /google/.cipd/pkgs/B9dhvBqFSHhs: invalid cross-device link

Error: building at STEP "RUN gclient sync": while running runtime: exit status 1
```

**Comment by caoccao** [comment]
It's been broken in docker for months. I think that's a docker issue. This [post](https://bbs.archlinux.org/viewtopic.php?id=241866) might work, but I haven't tested that because I always use physical box.

**Reply by josh-hemphill** [reply]
It looks like that just disables metacopy, and my docker and podman instances both have that disabled by default for compatibility.

**Reply by josh-hemphill** [reply]
If it is related to the filesystem drivers, then between my docker/podman on two machines and vm I've tried with both overlay and overlay2 drivers and both with and without native overlay diffs, none of that seems to affect it.

**Reply by caoccao** [reply]
I think the root cause is docker puts different layers in different mount points which breaks `gclient sync` when it tries to **rename** a temp folder after a successful download. Patching the host system or docker itself, to me, both serves the work.

**Reply by josh-hemphill** [reply]
I don't think there's anything I can patch; creating that file (it doesn't exist on my host or containers, and can't be created on my host) doesn't seem to affect anything. Reading more it sounds like it's an issue of how CoW interacts when trying to move files across mount boundaries. There's a overlayfs kernel option that allows full moving between mounts but I haven't been able find how to set that.

One of the points made everywhere I've seen talk about it is that it's usually the fault of the program running a rename/move and not falling back to copy-then-delete as other posix tools do when encountering an EXDEV error like this.

So I think I may have to go to the google chromium forum groups to see if they can help.

**Reply by caoccao** [reply]
I don't think Google will fix that. The practical way I'm thinking of is to hack the `cipd` from `mv` to `cp` + `rm`.

```sh
line 211: chmod +x "${CIPD_CLIENT_TMP}"
line 212: mv "${CIPD_CLIENT_TMP}" "${CLIENT}"
```

**Reply by josh-hemphill** [reply]
Won't that require cipd to be built from source since it's a binary?

**Reply by caoccao** [reply]
> Won't that require cipd to be built from source since it's a binary?

No, it is a shell script. I think an `sed` command before `gclient sync` might work. If it works, could you submit a PR to the dev branch?

**Reply by josh-hemphill** [reply]
Ah, I got confused since google also has a binary program of the same name written in `.go` that does something similar. I found the script.

**Reply by caoccao** [reply]
It's kind of a trash bin. Better try them all.

**Reply by josh-hemphill** [reply]
Yep, just tried doing sed on the script, and modifying it has no effect.
And just realized that the stack trace shows `client.go:1901`, so I think I'll need to look at the binary.

**Reply by josh-hemphill** [reply]
Got a response from a chromium person that the usage of the `rename()` function is used in cipd for its atomic guarantee (it doesn't use `mv`), so replacing it with `cp && mv` could introduce some subtle bugs. 

They suggested just combining the layers into a single script, which I'm starting to agree with since having a bunch of files next to each other but on different layers doesn't make a whole lot of sense except for when you're debugging. Like it would be nice to be able to start from the last layer, but having a script would also make it easier to use a single dockerfile for multiple platforms.

**Reply by josh-hemphill** [reply]
So I think I'm going to still open a bug with cipd in hopes someday there's a way for rename() to handle XDEV errors gracefully while maintaining atomicity. (looks like there's a flag option for the function for creating an intermediate file on overlay filesystems, which might be what I'm looking for, so there's hope)
In the meantime I'll look at writing some scripts and keep them as concise and straightforward as possible.

**Reply by caoccao** [reply]
Thank you for the update.

The challenge is the Linux and Android base images share the same layers till `fetch v8` and `install-build-deps.sh`. If all these commands are packaged in one script, the cache will not be reused. Why did I design it that way? Well, because I'm behind GFW. My VPN service charges me by data volume. 20GB per image is not a good thing for my wallet.

So, if someone could take care of the docker image build and push, I would be fine with a refactor of the docker files.

**Reply by josh-hemphill** [reply]
I just realized I could just copy>delete-origonal>move-back the whole `/google` directory so it's owned by the layer before running gclient sync.
I hate doing it, but in this case a bunch of extra time to the build isn't as bad as the download.

**Reply by caoccao** [reply]
Actually, you don't need to call `gclient sync` inside the docker if you performs `gclient sync` locally and copy everything over. But, that implies others are not able to build the images as you do.

**Reply by caoccao** [reply]
By the way, there's another issue #171 on redesigning the dockerfiles. I posted my suggestions some of which touch the Android image in that issue. You are welcome joining the discussion. Cheers 😄
